File: python/java_to_python/05-pgmAbstraccion/src/co/edu/campusucc/poo/Vehiculo.py
from abc import ABC, abstractmethod
import time


class Vehiculo(ABC):

    def __init__(self):
        self.marca = ""
        self.tipo = ""
        self.carroseria = ""
        self.velocidad = 0

    def __str__(self):
        return f"Abstract Class --> Vehiculo: {self.tipo} {self.carroseria} {self.marca}"

    @abstractmethod
    def startingEngine(self):
        pass

    @abstractmethod
    def stopingEngine(self):
        pass

    def speedUp(self):
        speedUpString = "speed up in Progress: "
        for i in range(self.velocidad):
            speedUpString += "⏩"

        for i in range(10):
            speedUpString += "⏩"
            print(speedUpString)
            time.sleep(1 - (i * 0.05))

        self.addSpeed(10)

    def addSpeed(self, velocidad):
        self.velocidad += velocidad
        return self.velocidad

    def downSpeed(self, velocidad):
        self.velocidad -= velocidad
        return self.velocidad

    def currentSpeed(self):
        print(f"Current Speed ⏩: {self.velocidad}")

    def parking(self):
        for i in range(5):
            print("¡¡🅿️🅿️ Parking in Progress ⚠️⚠️!!")
            time.sleep(1 - (i * 0.05))

    def breaking(self):
        speedUpString = ""
        for i in range(self.velocidad):
            speedUpString += "⏩"
        for i in range(10):
            speedUpString = speedUpString[:-1]
            print(f"¡¡🚥 Braking in Progress🚥: {speedUpString}")
            time.sleep(1 - (self.velocidad + i * 0.05))

        self.downSpeed(10)

    # Getters and Setters
    def getMarca(self):
        return self.marca

    def setMarca(self, marca):
        self.marca = marca

    def getTipo(self):
        return self.tipo

    def setTipo(self, tipo):
        self.tipo = tipo

    def getCarroseria(self):
        return self.carroseria

    def setCarroseria(self, carroseria):
        self.carroseria = carroseria

    def getVelocidad(self):
        return self.velocidad

    def setVelocidad(self, velocidad):
        self.velocidad = velocidad

File: ruby/pgms_ruby/05-pgmAbstraccion/Vehiculo.rb
require 'time'

class Vehiculo
  def initialize
    @marca = ""
    @tipo = ""
    @carroseria = ""
    @velocidad = 0
  end

  def to_s
    "Abstract Class --> Vehiculo: #{@tipo} #{@carroseria} #{@marca}"
  end

  def startingEngine
    raise NotImplementedError, "Subclass must implement abstract method"
  end

  def stopingEngine
    raise NotImplementedError, "Subclass must implement abstract method"
  end

  def speedUp
    speedUpString = "speed up in Progress: "
    @velocidad.times do
      speedUpString += "⏩"
    end

    10.times do |i|
      speedUpString += "⏩"
      puts speedUpString
      sleep(1 - (i * 0.05))
    end

    addSpeed(10)
  end

  def addSpeed(velocidad)
    @velocidad += velocidad
    return @velocidad
  end

  def downSpeed(velocidad)
    @velocidad -= velocidad
    return @velocidad
  end

  def currentSpeed
    puts "Current Speed ⏩: #{@velocidad}"
  end

  def parking
    5.times do |i|
      puts "¡¡🅿️🅿️ Parking in Progress ⚠️⚠️!!"
      sleep(1 - (i * 0.05).abs)
    end
  end

  def breaking
    speedUpString = ""
    @velocidad.times do
      speedUpString += "⏩"
    end

    10.times do |i|
      speedUpString = speedUpString[0..-2]
      puts "¡¡🚥 Braking in Progress🚥: #{speedUpString}"
      sleep((1 - (@velocidad + i * 0.05)).abs)
    end

    downSpeed(10)
  end

  # Getters and Setters
  def getMarca
    @marca
  end

  def setMarca(marca)
    @marca = marca
  end

  def getTipo
    @tipo
  end

  def setTipo(tipo)
    @tipo = tipo
  end

  def getCarroseria
    @carroseria
  end

  def setCarroseria(carroseria)
    @carroseria = carroseria
  end

  def getVelocidad
    @velocidad
  end

  def setVelocidad(velocidad)
    @velocidad = velocidad
  end
end